src/editor/mod.rs

```diff
@@ -65,6 +65,36 @@ impl Editor {
         editor
     }
 
+    pub fn cell_index(&self, x: u64, y: u64) -> Option<usize> {
+        let (width, height) = self.size;
+        if x >= width || y >= height {
+            None
+        }else{
+            Some((x + y * width) as usize)
+        }
+    }
+    
+    pub fn is_dirty_cell(&self, x: u64, y: u64) -> bool{
+        if let Some(idx) = self.cell_index(x, y) {
+            self.dirty[idx]
+        }else{
+            false
+        }
+    }
+
+    pub fn set_dirty_cell(&mut self, x: u64, y: u64) {
+        if let Some(idx) = self.cell_index(x, y) {
+            self.dirty[idx] = true;
+        }
+    }
+
+    fn rows<'a> (&'a self) -> Vec<&'a [GridCell]> {
```

Comment:
This is a way to preserve the row-by-row access pattern, without cloning the actual data. It just creates one slice per row, so the overhead should be pretty minimal.

Reply:
Looks awesome to me. See my top level comment